Remotion/Web/Core/WebString.cs

```diff
@@ -0,0 +1,208 @@
+    /// </summary>
+    /// <param name="left">The first value to be compared for equality.</param>
+    /// <param name="right">The second value to be compared for equality.</param>
+    /// <returns><see langword="true" /> if both <paramref name="left"/> and <paramref name="right"/> have the same string value
+    /// and the same type. Otherwise, <see langword="false" />.</returns>
+    public static bool operator == (WebString left, WebString right)
+    {
+      return left.Equals (right);
+    }
+
+    /// <summary>
+    /// Implements the inequality operator for <see cref="WebString"/>s. The operator is implemented the same way as the <see cref="Equals(WebString)"/> method.
+    /// </summary>
+    /// <param name="left">The first value to be compared for inequality.</param>
+    /// <param name="right">The second value to be compared for inequality.</param>
+    /// <returns><see langword="true" /> if both <paramref name="left"/> and <paramref name="right"/> are <see langword="null" /> do not have the same string value
+    /// or type. Otherwise, <see langword="false" />.</returns>
+    public static bool operator != (WebString left, WebString right)
+    {
+      return !left.Equals (right);
+    }
+
+    public static implicit operator WebString (string value) => new WebString (value, WebStringType.Text);
+
+    [CanBeNull]
+    private readonly string _value;
+    private readonly WebStringType _type;
+
+    /// <summary>
+    /// Gets the type of string represented by this <see cref="WebString"/>.
+    /// </summary>
+    /// <value>The type of string represented by this <see cref="WebString"/>.</value>
+    public WebStringType Type => _type;
+
+    /// <summary>
+    /// Tests whether the <see cref="WebString"/>s value is <see langword="null" /> or an empty string.
+    /// </summary>
+    public bool IsEmpty => string.IsNullOrEmpty (_value);
+
+    private WebString ([CanBeNull] string value, WebStringType type)
+    {
+      _value = value;
+      _type = type;
+    }
+
+    /// <summary>
+    /// Gets the raw string value of the <see cref="WebString"/> or an empty string in case the value is <see langword="null" />.
+    /// </summary>
+    [NotNull]
+    public string GetRawValue ()
+    {
+      return _value ?? string.Empty;
+    }
+
+    /// <summary>
+    /// Writes the <see cref="WebString"/>'s value to the specified <paramref name="writer"/>.
+    /// The value is escaped if necessary, depending on the <see cref="Type"/> of the <see cref="WebString"/>.
+    /// </summary>
+    /// <param name="writer">The <see cref="HtmlTextWriter"/> where the value will be appended to.</param>
+    /// <exception cref="InvalidOperationException">The <see cref="WebString"/>'s type is not compatible with the operation.</exception>
+    public void Write (HtmlTextWriter writer)
+    {
+      switch (_type)
+      {
+        case WebStringType.Text:
+          writer.Write (HttpUtility.HtmlEncode (GetRawValue()));
```

Comment:
I am quite lost with the second comment. Are you worried about `HtmlUtility.HtmlEncode` potentially allocating the string multiple times?

Since we are not in span-land and `TextWriter` does not support writing only a part of a string, the only problem I see is that the method might create unnecessary copies when a string contains multiple line endings.

Reply:
> the only problem I see is that the method might create unnecessary copies when a string contains multiple line endings.
Yes and now. The implementation in HtmlUtility does multplie string-search&replace, so creating substrings would be okay. there's also the (extremely questionable) alternative of rendering each char separately, but that would be future perf-optimization if we actually need it, given that most strings don't contain line breaks anyway.